x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/BinaryOperator.java

```diff
@@ -23,7 +22,9 @@ protected BinaryOperator(Location location, Expression left, Expression right, F
         super(location, left, right, function);
     }
 
-    protected abstract TypeResolution resolveInputType(DataType inputType);
+    protected abstract TypeResolution resolveInputTypeFirstArg(Expression e);
+
+    protected abstract TypeResolution resolveInputTypeSecondArg(Expression e);
```

Comment:
Most implementations seem to require the same type for the arguments and since the different between the two methods are just the `ParamOrdinal`, the previous `resolveType()` implementation might be kept while defining an abstract `protected abstract TypeResolution resolveInputType(Expression e, ParamOrdinal o)` which just pass the ordinal to the `Expressions` call.